frame/staking/src/pallet/impls.rs

```diff
@@ -191,9 +193,11 @@ impl<T: Config> Pallet<T> {
 				nominator_payout_count += 1;
 				let e = Event::<T>::Rewarded(nominator.who.clone(), imbalance.peek());
 				Self::deposit_event(e);
+				total_imbalance.subsume(imbalance);
 			}
 		}
 
+		T::Reward::on_unbalanced(total_imbalance);
```

Comment:
Note that even without this line the tests pass because in the mock setup we just have the default behavior for `on_unbalance`

Reply:
sounds like you should create a mock handler of on_unbalanced then to verify this specific behavior

Reply:
Added here: 2b6b63ea9f170d9d1d9263270ab8394dc28af5ea

Reply:
So let me try and see if I understand what was going on before...

There is a default `Drop` implementation for the `PositiveImbalance` type that gets returned from `make_payout`, and the `drop` behaviour is to call `on_unbalanced` defined somewhere, which was definitely not configurable via the `Reward` config item. Adding this line here shouldn't change anything at all behaviour-wise, because we're just surfacing the ability to configure the behaviour of the `on_unbalanced` method via `Reward`.

Now, regarding the `on_unbalanced` that was defined somewhere not `Reward`, I'd like to see where the code for that is, in order to convince myself that we are indeed not changing any behaviours here.

Reply:
- OnUnbalanced fallback impl: https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/frame/support/src/traits/tokens/imbalance/on_unbalanced.rs#L44-L46 which is implemented for `()` here https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/frame/support/src/traits/tokens/imbalance/on_unbalanced.rs#L55

- `PositiveImbalance` def: https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/frame/support/src/traits/tokens/currency.rs#L44
- `PositiveImbalanceOf` def: `type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId,>>::PositiveImbalance;`

Reply:
Okay, so `Reward` was set to `()`, that part I get, but what I don't get is how the new code introduced here is behaviorally equivalent to the previous one, since `on_balanced` was not called anywhere before the code changes.

Reply:
It is behaviorally equivalent in the scenario where the runtime was configured with `type Reward = ()` (which is what the polkadot runtimes do). This is because the behavior of `()` for `on_unbalanced` is the exact same as when the `Imbalance` is implicitly dropped. 

To concretely see what is happening in the Polkadot runtimes, we need to look at the currency implementation, which is the balances pallet:

https://github.com/paritytech/substrate/blob/6091c6ba54516b65071095b146beee796372c59f/frame/balances/src/lib.rs#L1387-L1392

In the new call path, we will try to call `try_drop` here, which will error if the imbalance is not zero and then we call `on_nonzero_unbalanced`

https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/frame/support/src/traits/tokens/imbalance/on_unbalanced.rs#L44-L46

And the implementation of `on_nonzero_unbalanced` is to call `drop`

https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/frame/support/src/traits/tokens/imbalance/on_unbalanced.rs#L50-L52

Reply:
Okay, so basically you're saying that the previous code relied on the `Drop` implementation of `PositiveImbalance` to ensure that the total issuance changed, and on the new code path that you've introduced, it doesn't explicitly rely on the `Drop` implementation, but it is dropped nonetheless after a call to `T::Reward::on_unbalanced`, _if_ the concrete type for `Reward` uses the default implementation of `on_unbalanced`.

Given that's the case, I think that it's then important to ensure that any type that gets assigned as `Reward` should have a sensible implementation of `on_unbalanced`, since if it doesn't call `drop` properly on the `Imbalance`, it may affect the total issuance of the currency, which is quite undesirable.

Some rust docs to the effect of **WARN** or **NOTE**  on the `on_unbalanced` method of the `OnUnbalanced` trait (or even on the trait itself) would suffice.

Reply:
Yeah I will look into adding some docs like that. The thing is though, there are use cases where `on_unbalanced` doesn't need to change the total issuance. For example it could just remove the equivalent funds from some treasury pot (and not change the total issuance).

Reply:
[c67fced](https://github.com/paritytech/substrate/pull/10887/commits/c67fced35319c2893c8669531f4925ac870606bb)

Reply:
So you added the docs on `RewardRemainder`, is that sufficient?

Reply:
🤦 meant `Reward` - fixed here [50a12a9](https://github.com/paritytech/substrate/pull/10887/commits/50a12a99df184b5afc3e99a121935063a2cde86f)

Reply:
But I think adding to `Reward` should be sufficient. Seems like adding them directly to the `OnUnbalanced` might give bad advice since we don't necessarily know all the ways people may be using it

Reply:
> But I think adding to `Reward` should be sufficient. Seems like adding them directly to the `OnUnbalanced` might give bad advice since we don't necessarily know all the ways people may be using it

Yeah, that looks fine to me.